### DISCUSSION TITLE
Trying to filter on specific Worksheets for User Adoption liveboards

### DISCUSSION THREAD

**Discussion by eckhami** [thoughtspot/cs_tools#179] [Q&A] [answered]
Hi, 
the User Adoption Liveboards which are deployed through CS Tools are already offering a lot of insight. 
However it is not possible to see, which Worksheet is being used by certain users, through Liveboards and Answers.

I have been thinking to create a view using TS_METADATA_OBJECT, TS_METADATA_COLUMN and TS_DEPENDENT_OBJECT.
The tricky part in my opinion is that for one row in TS_METADATA_OBJECT there are multiple entries in TS_METADATA_COLUMN (as each object can have one or more columns obviously) and for each row in TS_METADATA_COLUMN I would find multiple records in TS_DEPENDENT_OBJECT.

However, by using a distinct I was hoping to create a 1-1 relationship, as each Worksheet has a fixed set of columns and I just need to find the way from an answer to the worksheet (going through the COLUMN_GUID).
So I came up with below query, which should tell me, which worksheet is being used by a TS_METADATA_OBJECT:

```
select distinct
    ts_dependent_object.cluster_guid,
    ts_dependent_object.dependent_guid, --as answer_book_guid,
    ts_metadata_column.object_guid,     --as parent_guid,
    ts_metadata_object.name               as parent_name,
    ts_metadata_object.object_subtype     as parent_subtype
from ts_dependent_object
left join ts_metadata_column
on ts_dependent_object.cluster_guid = ts_metadata_column.cluster_guid
and ts_dependent_object.column_guid = ts_metadata_column.column_guid  
left join ts_metadata_object
on ts_metadata_column.cluster_guid = ts_metadata_object.cluster_guid
and ts_metadata_column.object_guid = ts_metadata_object.object_guid
```
this I joined to TS_METADATA_OBJECT through DEPENDENT_GUID to see the worksheet below. however I seem to miss something, as the results do not tie out.

Is there anything I missed?

Thanks
Michael

**Comment by boonhapus** [comment]
Hey @eckhami 

> However it is not possible to see, which Worksheet is being used by certain users, through Liveboards and Answers.

So this is definitely possible -- but it does come with some caveats. As you said, the relationships exist between `DEPENDENT > COLUMN > OBJECT`. A table is a collection of columns , and columns can be used in many different visualizations. So we find that there are many dependents (Answer, Liveboards, Worksheets) for a given Column, and there are many Columns for a given Object (Table, View, Worksheet).

![image](https://github.com/user-attachments/assets/f31d9905-2235-4480-ba84-eb061bfb3953)

The TS_BI_SERVER table contains a column called `ANSWER_BOOK_GUID` , which can be related to `DEPENDENT` on the `dependent_guid`. Really the dependents table is a subtype of `METADATA_OBJECT`. We've built it this way so it's simpler to consume.

Now, for the caveats. 😄 

> However it is not possible to see, which Worksheet is being used by certain users, through Liveboards and Answers.

`TS_BI_SERVER` as a data source (from the ThoughtSpot platform) does not attribute Users Actions to a Worksheet. It only attributes actions to Users on a given Visualization + Visualization "parent" (Answer/Liveboard). So while the Liveboard we offer in CS Tools shows Searches, which may be filtered by Worksheets.. what's really being shown is searches on an Answer (who has a parent Worksheet).

![image](https://github.com/user-attachments/assets/28222529-7c9b-41c4-aa23-0619c3d47dcc)

If I wanted to show the relationships between this information (Worksheet -> Dependents) , this is the kind of query I would write.

```sql
SELECT 
  "ta_1"."OBJECT_TYPE" "ca_1", 
  "ta_1"."NAME" "ca_2", 
  "ta_2"."COLUMN_NAME" "ca_3", 
  "ta_2"."IS_FORMULA" "ca_4", 
  count(DISTINCT "ta_3"."NAME") "ca_5"
FROM "SNOWFLAKE_DB"."CS_TOOLS_V150"."TS_DEPENDENT_OBJECT" "ta_3"
  RIGHT OUTER JOIN "SNOWFLAKE_DB"."CS_TOOLS_V150"."TS_METADATA_COLUMN" "ta_2"
    ON (
      "ta_3"."COLUMN_GUID" = "ta_2"."COLUMN_GUID"
      AND "ta_3"."CLUSTER_GUID" = "ta_2"."CLUSTER_GUID"
    )
  RIGHT OUTER JOIN "SNOWFLAKE_DB"."CS_TOOLS_V150"."TS_METADATA_OBJECT" "ta_1"
    ON (
      "ta_2"."CLUSTER_GUID" = "ta_1"."CLUSTER_GUID"
      AND "ta_2"."OBJECT_GUID" = "ta_1"."OBJECT_GUID"
    )
WHERE LOWER("ta_1"."NAME") = '(sample) retail - apparel'
GROUP BY 
  "ca_1", 
  "ca_2", 
  "ca_3", 
  "ca_4"
```

![image](https://github.com/user-attachments/assets/c1b21b31-457d-48a5-b64d-4ab20b79dcdf)

**Reply by boonhapus** [reply]
Ah -- one more thing. I didn't realize you asked about Users. Here's the above query with Users as well.

```sql
SELECT 
  "ta_1"."OBJECT_TYPE" "ca_1", 
  "ta_1"."NAME" "ca_2", 
  "ta_2"."COLUMN_NAME" "ca_3", 
  "ta_2"."IS_FORMULA" "ca_4", 
  count(DISTINCT "ta_3"."NAME") "ca_5", 
  count(DISTINCT "ta_4"."USER_ID") "ca_6"
FROM "SNOWFLAKE_DB"."CS_TOOLS_V150"."TS_BI_SERVER" "ta_4"
  LEFT OUTER JOIN "SNOWFLAKE_DB"."CS_TOOLS_V150"."TS_DEPENDENT_OBJECT" "ta_3"
    ON (
      "ta_4"."CLUSTER_GUID" = "ta_3"."CLUSTER_GUID"
      AND "ta_4"."ANSWER_BOOK_GUID" = "ta_3"."DEPENDENT_GUID"
    )
  RIGHT OUTER JOIN "SNOWFLAKE_DB"."CS_TOOLS_V150"."TS_METADATA_COLUMN" "ta_2"
    ON (
      "ta_3"."COLUMN_GUID" = "ta_2"."COLUMN_GUID"
      AND "ta_3"."CLUSTER_GUID" = "ta_2"."CLUSTER_GUID"
    )
  RIGHT OUTER JOIN "SNOWFLAKE_DB"."CS_TOOLS_V150"."TS_METADATA_OBJECT" "ta_1"
    ON (
      "ta_2"."CLUSTER_GUID" = "ta_1"."CLUSTER_GUID"
      AND "ta_2"."OBJECT_GUID" = "ta_1"."OBJECT_GUID"
    )
WHERE LOWER("ta_1"."NAME") = '(sample) retail - apparel'
GROUP BY 
  "ca_1", 
  "ca_2", 
  "ca_3", 
  "ca_4"
```

![image](https://github.com/user-attachments/assets/250496ff-7fb5-478d-ab0e-576b49e4f8a4)

**Comment by eckhami** [comment]
Thanks for all the information!
Unfortunately I did not yet have time to go through all of it yet, but there might be a slight misunderstanding though.
I am not so much interested in the worksheets being used, but rather I want to filter the User Adoption Liveboards on those data only, which are going back to certain Worksheets. Meaning to see all the Liveboards created, but filtered on specific underlying Worksheets...

I will still go through your code and see, if this helps me out.

Thanks

**Reply by boonhapus** [reply]
Makes sense @eckhami , this is exactly what the `Worksheet Column Utilization` worksheet is intended to solve.

![image](https://github.com/user-attachments/assets/15c8ca51-0c98-4c8d-83c3-35d9a4b7901b)


Here's the schema from that Worksheet.

![image](https://github.com/user-attachments/assets/b579a39b-4e61-44a9-884b-a436d8f43230)


Let me know how I can help!

**Comment by eckhami** [comment, marked as answer]
Hi,
with some help from the Thoughtspot team I found a possible solution to my question:
I need to be able to filter 'CS Tools ~ BI Server Advanced' to see results based on a specific subset of Worksheets.

First step was to see the relation of "Parent Worksheet" to Liveboards and answers associated with it. this was done through worksheet "CS Tools ~ Worksheet Column Utilisation"
![image](https://github.com/user-attachments/assets/eb14afdd-c174-4432-94e1-55b629c180c5)


From here I could extract the underlying SQL, which I then transferred into this code and saved as a view in ThoughtSpot:
```sql
SELECT 
  ta_2.cluster_guid,
  ta_2.dependent_guid,
  ta_1.object_guid,
  ta_1.NAME parent_name, --WORKSHEET
  CASE
    WHEN ta_2.OBJECT_TYPE = 'QUESTION_ANSWER_BOOK' THEN 'ANSWER'
    ELSE CASE
      WHEN ta_2.OBJECT_TYPE = 'PINBOARD_ANSWER_BOOK' THEN 'LIVEBOARD'
      ELSE null
    END
  END object_subtype, --DEPENDENT_TYPE, 
FROM TS_DEPENDENT_OBJECT ta_2
  RIGHT OUTER JOIN TS_METADATA_COLUMN MTA_0
    ON (
      ta_2.COLUMN_GUID = MTA_0.COLUMN_GUID
      AND ta_2.CLUSTER_GUID = MTA_0.CLUSTER_GUID
    )
  RIGHT OUTER JOIN TS_METADATA_OBJECT ta_1
    ON (
      MTA_0.CLUSTER_GUID = ta_1.CLUSTER_GUID
      AND MTA_0.OBJECT_GUID = ta_1.OBJECT_GUID
    )
WHERE (
  ta_2.object_type in ('QUESTION_ANSWER_BOOK', 'PINBOARD_ANSWER_BOOK') 
  AND (
    ta_1.OBJECT_TYPE = 'LOGICAL_TABLE'
    AND ta_1.OBJECT_SUBTYPE IN (
      'WORKSHEET', 'MODEL'
    )
  ) IN (TRUE)
)
GROUP BY all 
```

In (a copy of) "CS Tools ~ BI Server Advanced" I added the view with below Link definition:
![image](https://github.com/user-attachments/assets/357b9c2e-6bd6-48b3-94aa-6b342634b2d0)


Now I am able to filter the content of a Liveboard on specific Worksheets (in my case I called it "Parent Name"):
![image](https://github.com/user-attachments/assets/6ddcc9db-52cc-4808-a9e9-3fcee548f713)


Please let me know if I missed anything.
And thanks for all the help.

**Reply by boonhapus** [reply]
Excellent! Glad it worked out in the end @eckhami !
